File: PharmacyManagementAngular-ant/src/app/services/purchase-return.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { PurchaseReturnRequest, PurchaseReturnResponse, ApprovalStatus } from '../models/purchase-return.model';

@Injectable({
  providedIn: 'root',
})
export class PurchaseReturnService {

  private apiUrl = environment.apiUrl + 'purchase-returns';

  constructor(private http: HttpClient) {}

  getAllPurchaseReturns(): Observable<PurchaseReturnResponse[]> {
    return this.http.get<PurchaseReturnResponse[]>(this.apiUrl);
  }

  getPurchaseReturnById(id: number): Observable<PurchaseReturnResponse> {
    return this.http.get<PurchaseReturnResponse>(`${this.apiUrl}/${id}`);
  }

  getPurchaseReturnByNumber(returnNumber: string): Observable<PurchaseReturnResponse> {
    return this.http.get<PurchaseReturnResponse>(`${this.apiUrl}/number/${returnNumber}`);
  }

  getPurchaseReturnsBySupplier(supplierId: number): Observable<PurchaseReturnResponse[]> {
    return this.http.get<PurchaseReturnResponse[]>(`${this.apiUrl}/supplier/${supplierId}`);
  }

  getPurchaseReturnsByStatus(status: ApprovalStatus): Observable<PurchaseReturnResponse[]> {
    const params = new HttpParams().set('status', status);
    return this.http.get<PurchaseReturnResponse[]>(`${this.apiUrl}/status`, { params });
  }

  createPurchaseReturn(dto: PurchaseReturnRequest): Observable<PurchaseReturnResponse> {
    return this.http.post<PurchaseReturnResponse>(this.apiUrl, dto);
  }

  updatePurchaseReturn(id: number, dto: PurchaseReturnRequest): Observable<PurchaseReturnResponse> {
    return this.http.put<PurchaseReturnResponse>(`${this.apiUrl}/${id}`, dto);
  }

  approvePurchaseReturn(id: number): Observable<PurchaseReturnResponse> {
    const params = new HttpParams().set('status', ApprovalStatus.APPROVED);
    return this.http.patch<PurchaseReturnResponse>(`${this.apiUrl}/${id}/approve`, {}, { params });
  }

  cancelPurchaseReturn(id: number): Observable<PurchaseReturnResponse> {
    return this.http.delete<PurchaseReturnResponse>(`${this.apiUrl}/${id}`);
  }

  printPurchaseReturn(id: number): Observable<PurchaseReturnResponse> {
    return this.http.get<PurchaseReturnResponse>(`${this.apiUrl}/${id}/print`);
  }
}

File: PharmacyManagement/src/main/java/com/mishkat/PharmacyManagement/service/PurchaseReturnService.java
package com.mishkat.PharmacyManagement.service;

import com.mishkat.PharmacyManagement.dto.requestDTO.PurchaseReturnRequestDto;
import com.mishkat.PharmacyManagement.dto.responseDTO.PurchaseReturnResponseDto;

import java.util.List;

public interface PurchaseReturnService {
    PurchaseReturnResponseDto createPurchaseReturn(PurchaseReturnRequestDto dto);

    List<PurchaseReturnResponseDto> getAllPurchaseReturns();

    PurchaseReturnResponseDto getPurchaseReturnById(Long id);

    PurchaseReturnResponseDto getPurchaseReturnByNumber(String returnNumber);

    List<PurchaseReturnResponseDto> getReturnsBySupplierId(Long supplierId);

    PurchaseReturnResponseDto updatePurchaseReturn(Long id, PurchaseReturnRequestDto dto);

    void deletePurchaseReturn(Long id);
    PurchaseReturnResponseDto approvePurchaseReturn(Long id);
}
